src/slskd/Core/Data/Migrator.cs

```diff
@@ -0,0 +1,276 @@
+//     by the Free Software Foundation, either version 3 of the License, or
+//     (at your option) any later version.
+//
+//     This program is distributed in the hope that it will be useful,
+//     but WITHOUT ANY WARRANTY; without even the implied warranty of
+//     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//     GNU Affero General Public License for more details.
+//
+//     You should have received a copy of the GNU Affero General Public License
+//     along with this program.  If not, see https://www.gnu.org/licenses/.
+// </copyright>
+
+namespace slskd;
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using Serilog;
+using slskd.Migrations;
+
+/// <summary>
+///     Applies database migrations.
+/// </summary>
+public interface IMigration
+{
+    /// <summary>
+    ///     Determines whether the migration needs to be applied.
+    /// </summary>
+    /// <remarks>
+    ///     This method MUST be read-only and should contain no side effects.
+    /// </remarks>
+    /// <returns>A value indicating whether the migration needs to be applied.</returns>
+    bool NeedsToBeApplied();
+
+    /// <summary>
+    ///    Applies the migration.
+    /// </summary>
+    void Apply();
+}
+
+/// <summary>
+///     Applies database migrations.
+/// </summary>
+public class Migrator
+{
+    /// <summary>
+    ///     Initializes a new instance of the <see cref="Migrator"/> class.
+    /// </summary>
+    /// <param name="databases">A list of known databases.</param>
+    public Migrator(params string[] databases)
+    {
+        Databases = databases;
+    }
+
+    private string HistoryFileName { get; } = Path.Combine(Program.DataMigrationsDirectory, "history");
+    private IEnumerable<string> Databases { get; }
+    private ILogger Log { get; } = Serilog.Log.ForContext<Migrator>();
+
+    /// <summary>
+    ///     A map of all migrations, to be applied in the order they are specified (descending).
+    /// </summary>
+    private Dictionary<string, IMigration> Migrations { get; } = new()
+    {
+        { nameof(Z04012025_TransferStateMigration), new Z04012025_TransferStateMigration() },
+    };
+
+    /// <summary>
+    ///     Applies database migrations.
+    /// </summary>
+    /// <param name="force">Apply all migrations, regardless of whether there's evidence they have been applied already.</param>
+    public void Migrate(bool force = false)
+    {
+        Dictionary<string, DateTime> history = [];
+
+        Log.Information("Checking for outstanding database migrations...");
+
+        /*
+            load migration history from the history file in the root of the data directory. this file contains a key/value
+            pair for each migration that's been applied, along with the migration date.
+            if force=true, we're going to disregard history and overwrite it.
+        */
```

Comment:
suggestion: Store migration history in a database table like most migration systems (EF, Alembic, Liquibase, ...)

Reply:
i think i will; originally i didn't because Entity Framework would make it a hassle, but i don't have to use EF for this

Reply:
Since this topic came up, would it save you a lot of (ongoing) effort to use something like Flyway, or the above mentioned? I've used Flyway in a commercial setting and it worked fine, though it was heavy. I've heard https://github.com/golang-migrate/migrate is much better but I can't say I've tried it myself.

These sorts of tools are nice because you can usually stay away from your codebase and keep the separation of concerns very clean.